File: frontend/src/api/learning.ts
import { supabase } from '../lib/supabase';
import { callEdgeFunction } from '../lib/edgeFunctions';
import type {
  LearningEventRead,
  KnowledgeSnapshotRead,
} from './types';

export const learningService = {
  events: async (projectId: string, limit: number = 50): Promise<LearningEventRead[]> => {
    const { data, error } = await supabase
      .from('learning_event')
      .select('id, created_at, event_type, entity_type, entity_id, duration_ms, status, summary')
      .eq('project_id', projectId)
      .order('created_at', { ascending: false })
      .limit(limit);

    if (error) throw error;
    return (data ?? []) as LearningEventRead[];
  },

  snapshots: async (projectId: string, limit: number = 30): Promise<KnowledgeSnapshotRead[]> => {
    const { data, error } = await supabase
      .from('knowledge_snapshot')
      .select('*')
      .eq('project_id', projectId)
      .order('created_at', { ascending: false })
      .limit(limit);

    if (error) throw error;
    return (data ?? []) as KnowledgeSnapshotRead[];
  },

  status: async (projectId: string) =>
    callEdgeFunction('ai', { operation: 'learning-status', project_id: projectId }),

  rebuild: async (projectId: string) =>
    callEdgeFunction('ai', { operation: 'rebuild-learning', project_id: projectId }),
};

export { learningService as learningApi };

File: backend/src/api/routes/learning.py
from uuid import UUID
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional

from src.api.deps import get_db, get_project_or_404
from src.models.project import Project
from src.services import continuous_learning

router = APIRouter()


class LearningEventRead(BaseModel):
    id: str
    created_at: Optional[str] = None
    event_type: str
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    duration_ms: Optional[int] = None
    status: str
    summary: Optional[str] = None


class KnowledgeSnapshotRead(BaseModel):
    id: str
    created_at: Optional[str] = None
    total_trades: int
    total_patterns: int
    total_claims: int
    total_concepts: int
    total_sources: int
    total_similarities: int
    total_interpretations: int
    win_rate: float
    avg_rr: float
    expectancy: float
    knowledge_growth: float


class RebuildRequest(BaseModel):
    event_type: str = "manual_rebuild"


class RebuildResponse(BaseModel):
    event_id: str
    status: str
    duration_ms: int
    steps_completed: list[str]
    errors: list[str]


class LearningStatus(BaseModel):
    total_trades: int
    total_sources: int
    total_claims: int
    total_concepts: int
    total_interpretations: int
    total_patterns: int
    total_market_structures: int
    total_events: int
    last_event: Optional[dict] = None
    last_snapshot: Optional[dict] = None


@router.get("/events", response_model=list[LearningEventRead])
def get_learning_events(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
    limit: int = 50,
):
    events = continuous_learning.get_events(db, project_id=project_id, limit=limit)
    return [LearningEventRead(**e) for e in events]


@router.get("/snapshots", response_model=list[KnowledgeSnapshotRead])
def get_knowledge_snapshots(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
    limit: int = 30,
):
    snapshots = continuous_learning.get_snapshots(db, project_id=project_id, limit=limit)
    return [KnowledgeSnapshotRead(**s) for s in snapshots]


@router.post("/rebuild", response_model=RebuildResponse)
def rebuild_learning(
    project_id: UUID,
    request: RebuildRequest,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    result = continuous_learning.run_learning_pipeline(
        db, project_id=project_id, event_type=request.event_type,
    )
    return RebuildResponse(**result)


@router.get("/status", response_model=LearningStatus)
def get_learning_status(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    status = continuous_learning.get_status(db, project_id=project_id)
    return LearningStatus(**status)
